accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+| `VectorMask<T> Vector<T>.LessThanEqualsByVectorMask(Vector<T> v1, T val)`    |  
+
+In order to make use of the `VectorMask<T>` for conditional processing, we propose to extend the `Vector` operation API with overloaded methods to consume a `VectorMask`, e.g. `Vector<T>`:
+
+| Method  |
+| ------ | 
+| `VectorMask<T> Vector<T>.Add(Vector<T> v1, Vector<T> v2, VectorMask<T> cond)` | 
+| `VectorMask<T> Vector<T>.And(Vector<T> v1, Vector<T> v2, VectorMask<T> cond)` | 
+
+(OPEN: What API methods to extend here?)
+
+
+
+#### MaskExpr API
+
+`MaskExpr` is meant to start a small domain specific language (DSL) for Vector SIMD processing. To keep the implementation simple, we choose to directly use a `C#` expression (embedded DSL) encoded through a lambda. 
+
+In the following example:
+
+```C#
+public Vector256<int> SumVector256(ReadOnlySpan<int> source)
+{
+  Vector256<int> vresult = Vector256<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector256<int>.Count)
+  {
+    Vector256<int> v1 = new Vector256<int>(source.Slice(i));
+    VectorMask256<int> mask = v1.MaskExpr(x => x > 0 & x != 5);
+    vresult = Vector256.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+to JIT should logically create a `VectorMask` equivalent to `v1.GreaterThanByVectorMask(0) & v1.NotEqualByVectorMask(5)`.
+
+Given a usage of `MaskExpr`, `v.MaskExpr(x => e)`, where `v` represents a `Vector` and `e` represents an expression:
+
+- `v` defines the vector that we perform the conditional SIMD for (the implementation determines how the masking must be done)
+
+- `x` can be logically thought of as `v[i]` and is used to hint to the JIT on how to build the mask.
+
+- `e` defines an expression that the JIT will lift up in an appropriate masking. `e` must be a boolean expression, and is restricted to the standard boolean comparision operations.
+
+- If `e` does not type check to a boolean expression, and uses expressions other than boolean comparison operations and `x`, the JIT will throw `OperationNotSupportedError`.
+
+There are more advanced uses of a true "embedded DSL" and if the idea seems applicable, we can extend `MaskExpr` to handle many more interesting cases (at the cost of more advanced lifting from C# expression to something the JIT can work with).
+
+#### Leading and Trailing Processing API Methods
+
+We propose the following methods for each `Vector` API for processing leading and trailing elements. 
+
+| Method  |
+| ------ | 
+| `VectorMask<T> Vector<T>.CreateLeadingMask(int rem)` | 
+| `VectorMask<T> Vector<T>.LoadLeading(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.StoreLeading(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.CreateTrailingMask(int rem)` | 
+| `VectorMask<T> Vector<T>.LoadTrailing(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.StoreTrailing(Span<T> v1, VectorMask<T> mask)` | 
+
+### Internals Upgrades for EVEX/AVX512 Enabling
```

Comment:
This work should be coordinated with @JulieLeeMSFT. I would expect that Microsoft is providing code review and answering JIT design questions here. It would be great to confirm if the actual implementation is expected to be a collaborative effort or primarily driven by Intel.

Reply:
@kunalspathak is the most likely person to loop in for the register support.

There are a few people that semi-regularly work on the emitter including myself, @kunalspathak, and @EgorBo.

Someone needs to be looped in for the debugger work.

The VM work is expected to be small and likely doesn't need anyone dedicated, the exception being extending CG2 to support tracking the additional ISA flags (@davidwrighton).

Reply:
How `AVX512-VL` is tracked by the ISA flags might be of particular interest since its a single CPUID flag but is somewhat like the `x64` flag in that it impacts multiple ISAs